kubernetes-client-api/src/main/java/io/fabric8/kubernetes/client/utils/HttpClientUtils.java

```diff
@@ -136,16 +142,26 @@ public static String basicCredentials(String username, String password) {
     return "Basic " + encoded;
   }
 
+  private static final AtomicBoolean WARNED = new AtomicBoolean();
+
   public static HttpClient createHttpClient(Config config) {
-    // TODO: replace with service load
-    try {
-        return ((HttpClient.Factory) Class.forName("io.fabric8.kubernetes.client.okhttp.OkHttpClientFactory")
-                .getDeclaredConstructor()
-                .newInstance()).createHttpClient(config);
-    } catch (InstantiationException | IllegalAccessException | IllegalArgumentException | InvocationTargetException
-            | NoSuchMethodException | SecurityException | ClassNotFoundException e) {
-        throw KubernetesClientException.launderThrowable(e);
+    ServiceLoader<HttpClient.Factory> loader = ServiceLoader.load(HttpClient.Factory.class);
```

Comment:
This is something that will work flat class loaders, but may fail for other cases (e.g. OSGi, APT and so on).

I find it safer, to to perform SPI lookup using layers:
1. Plain `ServiceLoader.load(clazz)`
2. Using the current class classloader: `ServiceLoader.load(clazz, getClass().getClassLoader())`
3. Using TCCL: `ServiceLoader.load(clazz,  Thread.currentThread().getContextClassLoader())`

This is how we deal with adapters two: https://github.com/kameshsampath/kubernetes-client/blob/solo-extension/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/Adapters.java#L79

Reply:
1 and 3 are the same according to the ServiceLoader javadocs.  In adapters 2 is being called with a class from the extension.  Here it would be called with a class from the api.   Other than to guard against the case where someone has set an inappropriate TCCL, I don't see why you'd need 2 here.